elixir-sdk/native/sdk_core/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+use eppo_core::{
+    configuration_fetcher::{ConfigurationFetcher, ConfigurationFetcherConfig},
+    configuration_poller::{start_configuration_poller, ConfigurationPollerConfig},
+    configuration_store::ConfigurationStore,
+    eval::{Evaluator, EvaluatorConfig},
+    {Str, AttributeValue, Attributes},
+    ufc::{VariationType, Assignment},
+    SdkMetadata,
+    background::BackgroundThread,
+};
+
+use rustler::{Encoder, Env, NifResult, NifStruct, ResourceArc, Term, Atom};
+use rustler::types::atom;
+use std::{sync::{Arc, RwLock}};
+use std::panic::{RefUnwindSafe, UnwindSafe};
+use std::collections::HashMap;
+
+const SDK_METADATA: SdkMetadata = SdkMetadata {
+    name: "elixir",
+    version: env!("CARGO_PKG_VERSION"),
+};
+
+static CLIENT_INSTANCE: RwLock<Option<ResourceArc<EppoClient>>> = RwLock::new(None);
+
+#[derive(NifStruct)]
+#[module = "SdkCore.Config"]
+struct Config {
+    api_key: String,
+    base_url: String,
+    is_graceful_mode: bool,
+    poll_interval_seconds: Option<u64>,
+    poll_jitter_seconds: u64,
+}
+
+struct EppoClient {
+    evaluator: Evaluator,
+    background_thread: BackgroundThread,
+}
+
+#[rustler::resource_impl]
+impl rustler::Resource for EppoClient {}
+impl RefUnwindSafe for EppoClient {}
+impl UnwindSafe for EppoClient {}
+
+#[rustler::nif]
+fn init(config: Config) -> Result<(), String> {
+    // Validate config
+    if config.api_key.is_empty() {
+        return Err("Invalid value for api_key: cannot be blank".to_string());
+    }
+
+    let store = Arc::new(ConfigurationStore::new());
+    
+    let fetcher_config = ConfigurationFetcherConfig {
+        base_url: config.base_url.clone(),
+        api_key: config.api_key.clone(),
+        sdk_metadata: SDK_METADATA,
+    };
+
+    let fetcher = ConfigurationFetcher::new(fetcher_config);
+
+    let background_thread = BackgroundThread::start()
+        .map_err(|e| format!("Failed to start background thread: {}", e))?;
+
+    let poller_config = ConfigurationPollerConfig::new()
+        .with_interval(std::time::Duration::from_secs(
+            config.poll_interval_seconds.unwrap_or(30),
+        ))
+        .with_jitter(std::time::Duration::from_secs(config.poll_jitter_seconds));
+
+    let _poller = start_configuration_poller(
+        background_thread.runtime(),
+        fetcher,
+        store.clone(),
+        poller_config,
+    );
+
+
+    let evaluator = Evaluator::new(EvaluatorConfig {
+        configuration_store: store,
+        sdk_metadata: SDK_METADATA,
+    });
+
+    let client = ResourceArc::new(EppoClient {
+        evaluator,
+        background_thread,
+    });
+
+    // Set global instance
+    let mut instance = CLIENT_INSTANCE
+        .write()
+        .map_err(|e| format!("Failed to acquire write lock: {}", e))?;
+    
+    if let Some(existing) = instance.take() {
+        // Shutdown existing client
+        drop(existing);
+    }
+    
+    *instance = Some(client.clone());
```

Comment:
minor: if you can, I would like to push singleton up as much as possible. Ideally, having a non-singleton EppoClient "class" in Elixir with a thin wrapper to initialize/get instance.

This shall make it easier to support multiple clients in the future. Not sure about Elixir's threading model but that may also help us to get rid of locks (most of core is designed to allow concurrent access from multiple threads). Non-singletons are also kind of easier to reason about

Reply:
Hmm Elixir is functional and hence the singleton pattern does not make as much sense; hence I figured it would be easiest to completely hide the client object in Elixir. There's something called a GenServer we could try using, but I feel like it'd needlessly complicate things (and you can only have 1 with a given name, so wouldn't support multiple clients easily either). 

There's probably a good option available, I just don't know it

Reply:
Fixed this using a GenServer, actually is pretty nice, hope you like it!